feat: switch over to passport.js

coderbyheart opened the pull request:
This adds username/password based authentication,
using a cookie when accessing secure routes

See [./docs/authentication.md](https://github.com/distributeaid/shipment-tracker/blob/passport.js/docs/authentication.md) for description of the implementation.

- [x] use email instead of username
- [x] add email field
- [x] add password reset
- [x] confirm email after registration
- [x] send emails
- [x] change frontend `useAuth` hook to use context, so auth state can be access everywhere
- [x] add `ShipmentSendingHub` / `ShipmentReceivingHub` to db schema again
- [x] figure out what broke sequelize
- [x] add CNAME to Clever Cloud cellar bucket so app works properly (e.g. index.html is served as DirectoryIndex)
- [ ] send email with password token on password reset
- [ ] new auth forms need some love
- [ ] auto-refresh token on user activity
- [ ] convert clever-cloud infrastructure to code, using the [Clever Cloud CLI](https://www.clever-cloud.com/doc/getting-started/cli/)
- [ ] add CLI for making users admins

CAPTCHA will be handled in a separate PR: #335 

Includes changes to host it on [CleverCloud](https://www.clever-cloud.com/en/), so this PR can be reviewed on a working instance.

coderbyheart committed:
feat: switch over to passport.js

This adds username/password based authentication,
using a cookie when accessing secure routes

coderbyheart committed:
feat: update auth context and implement cookie renewal

coderbyheart committed:
fix: use uuid to generate usernames

coderbyheart committed:
fix: update docs

ghost commented:
### Changes to your CodeSee Architecture Map:
![View a CodeSee Map of these changes](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/distributeaid/shipment-tracker/314/f03c06d6/abbc3154f42106c3017a949e9672dc1da8a82681.svg)
#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

coderbyheart committed:
feat: add login and registration to frontend

coderbyheart committed:
fix: update depdendencies

coderbyheart committed:
fix: switch to Node.js 16

deammer commented on package.json:
```diff
@@ -19,38 +19,44 @@
-    "reflect-metadata": "^0.1.13",
-    "sequelize": "^6.6.2",
-    "sequelize-typescript": "^2.1.0"
+    "@sinclair/typebox": "0.20.5",
+    "@types/graphql-depth-limit": "1.1.2",
+    "@types/lodash": "4.14.172",
+    "@types/supertest": "2.0.11",
+    "ajv": "8.6.2",
+    "ajv-formats": "2.1.1",
+    "apollo-server": "3.3.0",
+    "apollo-server-express": "3.3.0",
+    "bcrypt": "5.0.1",
+    "body-parser": "1.19.0",
+    "compression": "1.7.4",
+    "cookie-parser": "1.4.5",
+    "cors": "2.8.5",
+    "dotenv": "10.0.0",
+    "express": "4.17.1",
+    "graphql": "15.5.1",
+    "graphql-depth-limit": "1.1.0",
+    "graphql-import-node": "0.0.4",
+    "graphql-scalars": "1.10.0",
+    "lodash": "4.17.21",
+    "nodemon": "2.0.12",
+    "passport": "0.4.1",
+    "passport-cookie": "1.0.9",
+    "pg": "8.7.1",
+    "prettier-plugin-organize-imports": "2.3.3",
+    "reflect-metadata": "0.1.13",
+    "sequelize": "6.6.2",
+    "sequelize-typescript": "2.1.0",
+    "supertest": "6.1.6",
+    "uuid": "8.3.2"
```
**praise:** thank you for pinning every dependency. I hope it wasn't too much of a pain.

coderbyheart replied:
Only a little, not all of them are on the latest release (e.g. sequelize-typescript)

deammer commented on src/authenticateRequest.ts:
```diff
@@ -1,169 +1,91 @@
-  [ROLES_KEY]: string[]
+type AuthCookiePayload = {
+  /** user ID */
+  i: number
+  /** user is admin */
+  a: boolean
+  /** user hash */
+  c: string
```
**question (non-blocking):** are those fields named that way to match the way `passport-cookie` formats its data?

coderbyheart replied:
No, but we encode this in the cookie sent to the browser, and there is a size limit, so I want to keep it as small as possible (while still using JSON).

deammer commented on src/models/user_account.ts:
```diff
@@ -25,8 +32,19 @@ export default class UserAccount extends Model<
+  @Unique
   @Column
-  public auth0Id!: string
+  public username!: string
```
**question (non-blocking):** why require a username instead of an email?

coderbyheart replied:
Yes, good point. I started this switch without emails, but I realize we need them anyway (for password reset), so I will remove the username field and turn it into an email field.

deammer commented on src/routes/login.ts:
```diff
@@ -0,0 +1,54 @@
+        username: valid.value.username,
+      },
+    })
+    if (user === null) {
+      // Penalize
+      await new Promise((resolve) => setTimeout(resolve, penaltySeconds * 1000))
+      return response.status(401).end()
+    }
+    if (!bcrypt.compareSync(valid.value.password, user.passwordHash)) {
+      // Penalize
+      await new Promise((resolve) => setTimeout(resolve, penaltySeconds * 1000))
+      return response.status(401).end()
+    }
```
**though (non-blocking):** this means that if I type the wrong username or password, I'll have to wait 10 seconds until I see an error message, right? I think that could become annoying from a UX perspective. I wonder if we could instead throttle access to that endpoint based on the user's IP (not necessarily in this PR though)

coderbyheart replied:
Yes, a more advanced feature of this would be remembering the user, however, the point here is to slow down attackers, which most likely will switch IP addresses. I think a better solution will be to implement a CAPTCHA, like https://friendlycaptcha.com/.

deammer replied:
I like the CAPTCHA solution 👍

coderbyheart committed:
fix: use email instead of username

coderbyheart commented:
@deammer I realize that I've build the authentication API outside of GraphQL, as REST. Do you feel that is an issue? They are pretty independent of each other, any way (and were before).

coderbyheart committed:
feat: prefer TypeScript migrations

deammer commented:
> @deammer I realize that I've build the authentication API outside of GraphQL, as REST. Do you feel that is an issue? They are pretty independent of each other, any way (and were before).

I feel like that's not a problem. The previous `/profile` endpoint was REST as well and we didn't have any issues.

coderbyheart committed:
fix: update dependencies

coderbyheart committed:
feat: implement password change (logged in)

coderbyheart committed:
fix: remove penalty

Will use CAPTCHA instead

coderbyheart committed:
feat: implement password reset with token

coderbyheart committed:
feat: make email lowercase

coderbyheart committed:
refactor: rewrite migrations in TypeScript

coderbyheart committed:
fix: email is lowercased

coderbyheart committed:
fix: ensure email is case-insensitive

coderbyheart committed:
feat: confirm registration using email and code

coderbyheart committed:
feat: add bus for emailing

coderbyheart committed:
feat: implement sending emails via SMTP

coderbyheart committed:
feat(frontend): hook up API

coderbyheart committed:
fix: use context in hook for shared state

coderbyheart committed:
fix: show which server is used for emails sending

coderbyheart committed:
fix: re-add ApolloProvider

coderbyheart committed:
fix: use token instead of code

coderbyheart committed:
fix(db): add migration for ShipmentReceivingHub and ShipmentSendingHub table

coderbyheart committed:
test(email): add test for appMailer

coderbyheart committed:
feat: implement FriendlyCaptcha

coderbyheart commented:
I've asked friendlyCaptcha if they would provide us with a free business license, because only that one uses EU-only endpoints.

coderbyheart committed:
test: add test for friendly captcha middleware

coderbyheart committed:
docs(backend): fix info in package.json

coderbyheart committed:
fix: convert CAPTCHA to use context

This allows to re-use the CAPTCHA solution over
different components

coderbyheart committed:
fix: remove friendlyCaptch for now

coderbyheart committed:
build: add backend-only server

coderbyheart committed:
build: add shortcut to build server and frontend

coderbyheart committed:
build(clevercloud): add DB configuration

coderbyheart committed:
fix(sequelize): allow multiple configs with URI

coderbyheart committed:
fix(frontend): remove friendly captcha for now

coderbyheart committed:
docs(frontend): set node.js version to 16

coderbyheart committed:
build: downgrade node version so it works on clever cloud

coderbyheart committed:
build: add script to deploy to Clever Cloud cellar

@see https://www.clever-cloud.com/blog/engineering/2020/06/24/deploy-cellar-s3-static-site/

coderbyheart committed:
build(clevercloud): install rclone

coderbyheart committed:
docs(frontend): node.js 16 is fine

coderbyheart committed:
build: use local rclone

coderbyheart committed:
build: use rclone from local folder

coderbyheart committed:
build: use only bucket name

coderbyheart committed:
build: echo progress

coderbyheart committed:
fix(frontend): only hardcode env vars for development

REACT_APP_VERSION is not used and maintained right now

coderbyheart committed:
fix(frontend): use only one server URL env var

coderbyheart committed:
fix: ensure Access-Control-Allow-Origin has no trailing slash

coderbyheart committed:
fix(passwords): do not enforce special characters

coderbyheart committed:
fix: sent cookies with sameSite=none

So Chrome will include it in cross-site requests

coderbyheart committed:
feat: remove UserProfileContext, add more auth pages

coderbyheart committed:
fix: include credentials with log-in, too

coderbyheart committed:
fix: remove unused LoginButton

coderbyheart committed:
fix(logout): reload to /

coderbyheart committed:
feat: include version string

coderbyheart committed:
build(frontend): fail on errors

coderbyheart committed:
ci(cd): change to correct folder

coderbyheart committed:
fix: print running version

coderbyheart committed:
fix: version middleware must come before routes

coderbyheart committed:
fix: logout url has colon

coderbyheart committed:
fix: remove log statement

coderbyheart commented:
I have to figure out the problem with sequelize before merging.

coderbyheart committed:
fix(env): remove dotenve so order of imports is not important

coderbyheart committed:
docs(devcontainer): update dev container to run tests

Update the DevContainer definition so users are able to run
`npm test`.

coderbyheart committed:
test(authentication): use correct cookie attributes
